tools/test/iree-benchmark-executable.mlir

```diff
@@ -0,0 +1,85 @@
+// Tests the iree-benchmark-executable tool against the portable VMVX target.
+// Other backends can be tested by using the appropriate compiler flags and
+// matching device and executable format flags.
+//
+// Examples:
+//   --iree-hal-target-backends=vmvx
+//     --device=local-sync or --device=local-task
+//     --executable_format=vmvx-bytecode-fb
+//   --iree-hal-target-backends=llvm-cpu
+//     --device=local-sync or --device=local-task
+//     --executable_format=embedded-elf-x86_64
+//     --executable_format=system-dll-x86_64
+//   --iree-hal-target-backends=vulkan-spirv
+//     --device=vulkan
+//     --executable_format=vulkan-spirv-fb
+
+// RUN: iree-compile \
+// RUN:     --compile-mode=hal-executable \
+// RUN:     --iree-hal-target-backends=vmvx \
+// RUN:     %s | \
+// RUN: iree-benchmark-executable \
+// RUN:     --device=local-sync \
+// RUN:     --executable_format=vmvx-bytecode-fb \
+// RUN:     --executable_file=- \
+// RUN:     --entry_point=0 \
+// RUN:     --binding=512xf32 \
+// RUN:     --binding=512xf32 \
+// RUN:     --binding=512xf32 \
+// RUN:     --workgroup_count=1,1,1 \
+// RUN:     --workgroup_count=512,1,1 | \
+// RUN: FileCheck %s --check-prefix=CHECK-VMVX
+
+// CHECK-VMVX: BM_dispatch_1x1x1
+// CHECK-VMVX: BM_dispatch_512x1x1
+
+// lhs * rhs => dst / s0b0 * s0b1 => s0b2
+#pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [
+  #hal.descriptor_set.layout<0, bindings = [
+    #hal.descriptor_set.binding<0, storage_buffer>,
+    #hal.descriptor_set.binding<1, storage_buffer>,
+    #hal.descriptor_set.binding<2, storage_buffer>
+  ]>
+]>
+hal.executable.source public @executable {
+  hal.executable.export public @elementwise_mul ordinal(0) layout(#pipeline_layout) attributes {
+    workgroup_size = [1 : index, 1 : index, 1 : index]
+  } {
+  ^bb0(%device: !hal.device):
+    // Unused - the workgroup count is provided to the tool.
+    %c1 = arith.constant 1 : index
+    hal.return %c1, %c1, %c1 : index, index, index
+  }
+  builtin.module {
+    func.func @elementwise_mul() {
+      %lhs = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(32) : !flow.dispatch.tensor<readonly:tensor<4xf32>>
+      %rhs = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(32) : !flow.dispatch.tensor<readonly:tensor<4xf32>>
+      %dst = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(32) : !flow.dispatch.tensor<writeonly:tensor<4xf32>>
+      // TODO(#16554): GPU/SPIR-V lowering doesn't handle workgroup size queries.
+      // %workgroup_size_x = hal.interface.workgroup.size[0] : index
+      %workgroup_size_x = arith.constant 1 : index
+      %workgroup_id_x = hal.interface.workgroup.id[0] : index
+      %workgroup_count_x = hal.interface.workgroup.count[0] : index
```

Comment:
Do you want all of this IR in the test for the tool? I could also see a test using an input dialect program (e.g. simple_abs, or a linalg matmul) together with `--iree-hal-dump-executable-binaries-to`.

Reply:
Yes - the ABI is required to be materialized here - anything we generate from earlier forms has an unspecified ABI and there's no way to then construct the flags to execute it.